src/machine/aot/aot.x64.c

```diff
@@ -0,0 +1,1283 @@
+      }
+      break;
+    case AOT_TAG_X64_REGISTER:
+      | mov rdx, Rq(bits.value.x64_reg)
+      | mov ecx, 64
+      | and edx, 0x3F
+      | sub ecx, edx
+      | shl rax, cl
+      | mov ecx, edx
+      if (is_signed) {
+        | sar rax, cl
+      } else {
+        | shr rax, cl
+      }
+      break;
+  }
+
+  | op2_r_x mov, target, rax
+
+  return DASM_S_OK;
+}
+
+int aot_exit(AotContext* context, int code)
+{
+  dasm_State** Dst = &context->d;
+  | mov rax, code
+  | jmp ->exit
+  return DASM_S_OK;
+}
+
+int aot_add_cycles(AotContext* context, uint64_t cycles)
+{
+  int ret;
+  dasm_State** Dst = &context->d;
+  if (cycles == 0) {
+    return DASM_S_OK;
+  }
+  | load_imm rax, cycles
+  | add machine->cycles, rax
+  | mov rax, machine->max_cycles
+  | cmp machine->cycles, rax
+  | jna >1
+  ret = aot_exit(context, CKB_VM_ASM_RET_MAX_CYCLES_EXCEEDED);
+  if (ret != DASM_S_OK) { return ret; }
+  |1:
+  return DASM_S_OK;
+}
+
+int aot_ecall(AotContext* context)
+{
+  return aot_exit(context, CKB_VM_ASM_RET_ECALL);
+}
+
+int aot_ebreak(AotContext* context)
+{
+  return aot_exit(context, CKB_VM_ASM_RET_EBREAK);
+}
+
+int aot_mov_pc_internal(AotContext* context, AotValue value)
+{
+  int ret;
+  dasm_State** Dst = &context->d;
+
+  switch (value.tag) {
+    case AOT_TAG_REGISTER:
+      /*
+       * At encoding time we cannot tell what address to jump to here,
+       * so all we can do here is to update the correct PC register,
+       * exit from current function call and defer to the machine to
+       * handle this.
+       */
+      ret = aot_mov_x64(context, X64_RCX, value);
+      if (ret != DASM_S_OK) { return ret; }
+      | mov machine->pc, rcx
+      ret = aot_exit(context, CKB_VM_ASM_RET_DYNAMIC_JUMP);
+      if (ret != DASM_S_OK) { return ret; }
+      break;
+    case AOT_TAG_IMMEDIATE:
+      /*
+       * It's very unlikely we will expand CKB-VM to have more than 4GB memory,
+       * hence we are leveraging this fact to encode dynasm dynamic label into
+       * the upper 32-bit part of immediate. This way we can both update correct
+       * PC value, and do quick jumps.
+       * Also, since the maximum number of labels supported now is 65535, we
+       * actually only need 16 bits of the upper 32-bit part, hence we are using
+       * the highest byte to store flags for different kinds of labels.
+       */
+      switch ((uint8_t) (value.value.i >> 56)) {
+        case 0x80:
+          /*
+           * This means just writing the result without actually jump
+           */
+          | load_imm rcx, (value.value.i & 0xFFFFFFFFFFFFFF)
+          | mov qword machine->pc, rcx
+          break;
+        case 0x40:
+          | mov qword machine->pc, ((uint32_t)(value.value.i & 0x7FFFFFFF))
+          | jmp =>((value.value.i >> 32) ^ 0x40000000)
+          break;
+        case 0x0:
+          | load_imm rcx, value.value.i
+          | mov machine->pc, rcx
+          ret = aot_exit(context, CKB_VM_ASM_RET_DYNAMIC_JUMP);
```

Comment:
Is this an unexpected jump in a code block, so we return an error here?

Reply:
I suggest we consider the return here as an indicator that the current x64 code doesn't know where to go next, hence it would exit first , and use Rust code's aid to figure out the next instruction to execute, the code here does exactly that. This could due to a jump to a place where no label is generated, or a jump based on a runtime value in a register

Reply:
I see. I thought it returns the code as the final VM script result, indeed the exit code is passed to the `match result` block in `AsmMachine::run`.